core/src/apps/zcash/unified_addresses.py

```diff
@@ -0,0 +1,118 @@
+"""
+Implementation of encoding and decoding of Zcash
+unified addresses according to the ZIP-316.
+see: https://zips.z.cash/zip-0316
+"""
+
+from typing import Dict
+
+from trezor.crypto.bech32 import Encoding, bech32_decode, bech32_encode, convertbits
+from trezor.enums import ZcashReceiverTypecode as Typecode
+from trezor.utils import BufferReader, empty_bytearray
+from trezor.wire import DataError
+
+from apps.common.coininfo import CoinInfo
+from apps.common.readers import read_compact_size
+from apps.common.writers import write_bytes_fixed, write_compact_size
+
+from .f4jumble import f4jumble, f4unjumble
+
+
+def receiver_length(typecode: int):
+    """Byte length of a receiver."""
+    if typecode in [Typecode.P2PKH, Typecode.P2SH]:
+        return 20
+    if typecode in [Typecode.SAPLING, Typecode.ORCHARD]:
+        return 43
+    raise ValueError
+
+
+def prefix(coin: CoinInfo):
+    """Prefix for a unified address."""
+    if coin.coin_name == "Zcash":
+        return "u"
+    if coin.coin_name == "Zcash Testnet":
+        return "utest"
+    raise ValueError
+
+
+def padding(hrp: str) -> bytes:
+    assert len(hrp) <= 16
+    return hrp.encode() + bytes(16 - len(hrp))
+
+
+def encode(receivers: Dict[Typecode, bytes], coin: CoinInfo) -> str:
+    # multiple transparent receivers forbidden
+    assert not (Typecode.P2PKH in receivers and Typecode.P2SH in receivers)
```

Comment:
Also
```python
assert Typecode.SAPLING in receivers or Typecode.ORCHARD in receivers
```
But I am wondering whether it is right to treat this with `assert`. It would seem more appropriate to use `if ... raise` at least until we have some guarantee that the calling code is checking this. (Right now `encode()` is not used.)

Reply:
`assert` seems to fine to me since input of this function will be always created internally.

Reply:
[6a410a8](https://github.com/trezor/trezor-firmware/pull/2398/commits/6a410a833c9ba28c91a0b364a7856b7fd08a5cbf)